File: 126. Word Ladder II.py
# https://leetcode.com/problems/word-ladder-ii/

# A transformation sequence from word beginWord to word endWord using a
# dictionary wordList is a sequence of words beginWord -> s1 -> s2 -> ... -> sk
# such that:

# Given two words, beginWord and endWord, and a dictionary wordList, return all
# the shortest transformation sequences from beginWord to endWord, or an empty
# list if no such sequence exists. Each sequence should be returned as a list
# of the words [beginWord, s1, s2, ..., sk].

###############################################################################

# bfs to find the shortest path length
# dfs to get all the pathes

from collections import defaultdict, deque
class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        if not beginWord or not endWord or not wordList or endWord not in wordList or beginWord == endWord: 
            return []

        # construct next words dict for dfs
        # key: word, value = list of next words
        next_words_dict = defaultdict(set)

        # construct unused words set, 
        words_avail = set(wordList)
        if beginWord in words_avail: #remove beginWord
            words_avail.remove(beginWord)
        
        # bfs to find the shortest path
        queue = deque([beginWord])
        path_exist = False
        while queue and not path_exist:
            words_used_this_layer = set() # reset at every layer
            for _ in range(len(queue)): # scan by layers
                word = queue.popleft()
                
                found = False # reset for every word
                for i in range(len(word)): # alter every char of word
                    if found: break # stop when one child reach endWord
                    
                    for char in 'abcdefghijklmnopqrstuvwxyz':
                        if char != word[i]: # must differ
                            next_word = word[:i] + char + word[i+1:]
                            
                            if next_word in words_avail:
                                queue.append(next_word) # add to queue
                                next_words_dict[word].add(next_word) # add to next_words_dict
                                words_used_this_layer.add(next_word)
                                
                                if next_word == endWord:
                                    found = True
                                    path_exist = True
                                    break # stop generating child for this word

            words_avail -= words_used_this_layer # remove all used word at the end
            
        if not path_exist: return []
        
        # dfs to find all shortest path
        self.ans = []
        self.holder = [beginWord]
        self.endWord = endWord
        self.next_words_dict = next_words_dict
        
        self.dfs()
        return self.ans
    
    def dfs(self):
        if self.holder[-1] == self.endWord: # exit case
            self.ans.append(self.holder.copy())
            return
        
        word = self.holder[-1]
        for next_word in self.next_words_dict[word]:
            self.holder.append(next_word)
            self.dfs()
            self.holder.pop()

File: 126. Word Ladder II.cpp
// https://leetcode.com/problems/word-ladder-ii/

// A transformation sequence from word beginWord to word endWord using a dictionary
// wordList is a sequence of words beginWord -> s1 -> s2 -> ... -> sk such that:

// Given two words, beginWord and endWord, and a dictionary wordList, return all
// the shortest transformation sequences from beginWord to endWord, or an empty
// list if no such sequence exists. Each sequence should be returned as a list of
// the words [beginWord, s1, s2, ..., sk].

////////////////////////////////////////////////////////////////////////////////

// bfs to find length, dfs to find all sequences
class Solution {
public:
    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> availWords;
        for (const string& word : wordList) availWords.insert(word);
        availWords.erase(beginWord);
        if (availWords.find(endWord) == availWords.end()) return {};
        
        queue<string> todo; todo.push(beginWord);
        bool found = false;
        while(!todo.empty() and !found) {
            int nCurrLayer = todo.size();
            unordered_set<string> usedWordsCurrLayer;
            while (nCurrLayer--) {
                string currWord = todo.front(); todo.pop();
                bool foundByCurrWord = false; // for early stopping
                for (int i = 0; i < currWord.size(); ++i) { // change every index
                    if (foundByCurrWord) break;
                    string nextWord = currWord;
                    for (int j = 0; j < 26; ++j) { // change currWord[i] to 'a'--'z'
                        nextWord[i] = 'a' + j;
                        if (nextWord[i] == currWord[i]) continue;
                        if (availWords.find(nextWord) != availWords.end()) { // legal child
                            usedWordsCurrLayer.insert(nextWord);
                            wordsMap[currWord].insert(nextWord); // add to wordsMap
                            todo.push(nextWord);
                            if (nextWord == endWord) { 
                                foundByCurrWord = true; found = true;
                                break;
                            }
                        }
                    }
                }
            }
            // remove used words from availWords
            for (const string& word : usedWordsCurrLayer) availWords.erase(word);
        }
        if (!found) return {};
        // dfs to find all paths
        holder.push_back(beginWord);
        dfs(endWord);
        return ans;
    }
private:
    // wordsMap[word] = set of next words, useful for dfs
    unordered_map<string, unordered_set<string>> wordsMap;
    vector<string> holder;
    vector<vector<string>> ans;
    void dfs(const string& endWord) {
        if (holder.back() == endWord) {
            ans.push_back(holder); return;
        }
        string currWord = holder.back();
        for (const auto& nextWord : wordsMap[currWord]) {
            holder.push_back(nextWord);
            dfs(endWord);
            holder.pop_back();
        }
        return;
    }
};
